File: frontend/src/types/monthlyBilling.ts
export type MonthlyRateStatus = "MATCHED" | "MISSING" | "AMBIGUOUS";
export type RegisterStatus = "Pending" | "Partially Paid" | "Paid" | "Cancelled";
export interface MonthStatus { month:string; status:"DRAFT"|"FINALIZED"|"REOPENED"; revision_number:number; finalized_at:string|null; reopened_at:string|null; notes:string|null }
export interface ExecutiveMonthlyBilling {
  executive_id: number | null; executive: string; rate: string | null; rate_display: string;
  bank_counts: Record<string, number>; total_points: number; gross_payment: string | null;
  advance: string; net_payment: string | null; paid: string; balance: string | null;
  payment_status: RegisterStatus; rate_status: MonthlyRateStatus; register_id: number | null; is_finalized: boolean;
  payment_date?:string|null; payment_reference?:string|null; remarks?:string|null; snapshot_revision?:number|null;
}
export interface BankMonthlyBilling {
  case_id: number|null; date: string; bank: string | null; los_no: string | null; name: string | null;
  address: string | null; city: string | null; mobile: string | null; status: string;
  remark: string | null; rate: string | null; rate_status: MonthlyRateStatus;
}
export interface MonthlyBillingResponse {
  month: string; executive_billing: ExecutiveMonthlyBilling[]; bank_billing: BankMonthlyBilling[];
  summary: { total_cases: number; billable_cases: number; missing_executive_rates: number;
    missing_bank_rates: number; ambiguous_rates: number; total_executive_payment: string; total_bank_billing: string };
  month_status: MonthStatus;
}
export interface BankPayment { id:number; billing_month:string; bank:string; city:string; billed_amount:string; received_amount:string; balance_amount:string; status:RegisterStatus; payment_date:string|null; payment_reference:string|null; remarks:string|null; is_finalized:boolean }
export interface BillingDashboard { month:string; month_status:MonthStatus; total_bank_billing:string; bank_received:string; bank_outstanding:string; total_executive_payout:string; executive_paid:string; executive_outstanding:string; expected_gross_margin:string; realized_cash_margin:string; bank_summary:BankPayment[]; executive_summary:ExecutiveMonthlyBilling[] }
export interface PaymentRegisterPayload {
  billing_month: string; executive_id: number; advance_amount: number; paid_amount: number;
  payment_date?: string | null; payment_reference?: string | null; remarks?: string | null;
  finalize?: boolean; regenerate?: boolean;
}

File: backend/app/schemas/monthly_billing.py
from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

RateStatus = Literal["MATCHED", "MISSING", "AMBIGUOUS"]
RegisterStatus = Literal["Pending", "Partially Paid", "Paid", "Cancelled"]


class ExecutiveMonthlyRow(BaseModel):
    executive_id: int | None
    executive: str
    rate: Decimal | None
    rate_display: str
    bank_counts: dict[str, int]
    total_points: int
    gross_payment: Decimal | None
    advance: Decimal
    net_payment: Decimal | None
    paid: Decimal
    balance: Decimal | None
    payment_status: RegisterStatus
    rate_status: RateStatus
    register_id: int | None
    is_finalized: bool
    payment_date: date | None = None
    payment_reference: str | None = None
    remarks: str | None = None
    snapshot_revision: int | None = None


class BankMonthlyRow(BaseModel):
    case_id: int | None
    visit_id: int | None = None
    visit_type: str | None = None
    date: date
    company: str | None
    bank: str | None
    los_no: str | None
    name: str | None
    address: str | None
    city: str | None
    district: str | None
    mobile: str | None
    status: str
    remark: str | None
    rate: Decimal | None
    rate_status: RateStatus
    bank_rate_id: int | None = None


class MonthlySummary(BaseModel):
    total_cases: int
    billable_cases: int
    missing_executive_rates: int
    missing_bank_rates: int
    ambiguous_rates: int
    total_executive_payment: Decimal
    total_bank_billing: Decimal


class MonthlyBillingResponse(BaseModel):
    month: str
    executive_billing: list[ExecutiveMonthlyRow]
    bank_billing: list[BankMonthlyRow]
    summary: MonthlySummary
    month_status: "MonthStatusResponse"


class PaymentRegisterUpdate(BaseModel):
    billing_month: str = Field(pattern=r"^\d{4}-(0[1-9]|1[0-2])$")
    executive_id: int
    advance_amount: Decimal = Field(default=Decimal("0"), ge=0)
    paid_amount: Decimal = Field(default=Decimal("0"), ge=0)
    payment_date: date | None = None
    payment_reference: str | None = None
    remarks: str | None = None
    finalize: bool = False
    regenerate: bool = False


class PaymentRegisterResponse(BaseModel):
    id: int
    billing_month: date
    executive_id: int
    executive: str
    gross_payment: Decimal
    advance_amount: Decimal
    net_payment: Decimal
    paid_amount: Decimal
    balance_amount: Decimal
    status: RegisterStatus
    payment_date: date | None
    payment_reference: str | None
    remarks: str | None
    is_finalized: bool
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)


class MonthStatusResponse(BaseModel):
    month: str
    status: Literal["DRAFT", "FINALIZED", "REOPENED"]
    revision_number: int = 0
    finalized_at: datetime | None = None
    reopened_at: datetime | None = None
    notes: str | None = None


class FinalizeMonthRequest(BaseModel):
    month: str = Field(pattern=r"^\d{4}-(0[1-9]|1[0-2])$")
    notes: str | None = None


class ReopenMonthRequest(BaseModel):
    month: str = Field(pattern=r"^\d{4}-(0[1-9]|1[0-2])$")
    reason: str = Field(min_length=1)


class RegenerateMonthRequest(BaseModel):
    month: str = Field(pattern=r"^\d{4}-(0[1-9]|1[0-2])$")
    confirm: bool


class BankPaymentUpdate(BaseModel):
    billing_month: str = Field(pattern=r"^\d{4}-(0[1-9]|1[0-2])$")
    bank: str
    company: str = ""
    district: str = ""
    city: str = ""
    received_amount: Decimal = Field(ge=0)
    status: RegisterStatus | None = None
    payment_date: date | None = None
    payment_reference: str | None = None
    remarks: str | None = None


class BankPaymentResponse(BaseModel):
    id: int
    billing_month: date
    bank: str
    company: str
    district: str
    city: str
    billed_amount: Decimal
    received_amount: Decimal
    balance_amount: Decimal
    status: RegisterStatus
    payment_date: date | None
    payment_reference: str | None
    remarks: str | None
    is_finalized: bool
    model_config = ConfigDict(from_attributes=True)


class BillingDashboardResponse(BaseModel):
    month: str
    month_status: MonthStatusResponse
    total_bank_billing: Decimal
    bank_received: Decimal
    bank_outstanding: Decimal
    total_executive_payout: Decimal
    executive_paid: Decimal
    executive_outstanding: Decimal
    expected_gross_margin: Decimal
    realized_cash_margin: Decimal
    bank_summary: list[BankPaymentResponse]
    executive_summary: list[ExecutiveMonthlyRow]


MonthlyBillingResponse.model_rebuild()
